Source/tmModel/wnlib/low/wnasrt.man
NAME
     wnasrt -- allows application to commit suicide

SYNOPSIS
     #include "wnasrt.h"

     wn_assert(bool condition)

     wn_assert_notreached()

     wn_assert_warn(bool condition)

     wn_assert_warn_notreached()

     void wn_set_assert_print(passert_print)
     void (*passert_print)(char string[]);

     void wn_default_assert_print(char string[]);

     void wn_set_assert_crash(passert_crash)
     void (*passert_crash)(void);

     void wn_default_assert_crash(void);

DESCRIPTION
     "wn_assert" evaluates "condition".  If "condition" is TRUE, execution
     continues.  If "condition" is FALSE, it prints a message giving the
     file name and line number where it was called from.  Then it crashes
     the application using abort(3).

     "wn_assert_notreached" prints a message giving the file name and 
     line number where it was called from.  Then it crashes the
     application using abort(3).

     "wn_assert_warn" evaluates "condition".  If "condition" is FALSE, it 
     prints a message giving the file name and line number where it was
     called from.

     "wn_assert_warn_notreached" prints a message giving the file name and 
     line number where it was called from.

     The above 4 routines may also print messages indicating function name
     and in the case of asserts, the expression that failed.  This
     functionality may not be available on all platforms.

     "wn_set_assert_print" allows the programmer to change the
     routine used for printing messages.  The default is to print
     to standard error, as specified in "wn_default_assert_print".

     "wn_set_assert_crash" allows the programmer to change the
     routine used for crashing, or perhaps to a routine that does not
     crash (does a longjump, for example).  The default is to crash
     the application using abort(3), as specified in "wn_default_assert_crash".

     wn_assert*() are expressions, not statements.  They are of type void.
     They can be used like statements, or used within expressions, but they
     don't return a value.  It is anticipated that their being able to be
     used in expressions could come in useful in complex macros.  Otherwise
     they can be treated syntactically like subroutine calls of type void.

     wn_assert*() calls can be overridden with environment variables.  If
     the appropriate override is set, a given assert will not bomb, and it
     will not print the usual error message.  The first time any suppressed
     assert or notreached is triggered, a small warning is produced, later
     calls are all silent.

     To override an assert or notreached, get the filename and line number,
     and run the xlateasrt program in wnlib/acc/low with

	xlateasrt -override_code <filename> <line number>

     Set and export the given environment variable, and rerun the program,
     the assert / notreached will be suppressed.

EXAMPLE

     The user may say

	wn_assert(i == 2);

     somewhere in his code.  If i != 2 when this statement gets evaluated,
     the assert fails.  A message something like this is printed

	------------------------------------------------------------
			       HELP!!!
	This program has encountered a severe internal error.
	Please report the following message to the
	program's developers so that they can fix the problem,
	an override may be available:
	------------------------------------------------------------
	Assertion botched -- forcing crash
	Expr (i == 2) was false
	in foo_routine() at line 57 in file barfile.c
        Aborted

     and the program aborts.

     If, upon examining the code, a decision is made that the assert should
     be ignored, it can be overridden.  To get the environment variable
     that must be set, run xlateasrt in wnlib/acc/low with -override_code,
     giving it the filename and line number from the message

	% xlateasrt -override_code barfile.c 57

     and it will give you a long name of an environment variable, in this case

	Override string == "WN_ASSERT_SUPPRESS_barfile_c_57_91120e3f"

     Set and export this variable.  On bash and sh, say

	% WN_ASSERT_SUPPRESS_barfile_c_57_91120e3f=
	% export WN_ASSERT_SUPPRESS_barfile_c_57_91120e3f

     On csh and tcsh, say

	% setenv WN_ASSERT_SUPPRESS_barfile_c_57_91120e3f

     Then run your program again, when it encounters the problem, you will
     get a brief message

	Warning: at least one assert suppressed, by environment variable
	    "WN_ASSERT_SUPPRESS_barfile_c_57_91120e3f"

     but nothing else, your program will continue on, and subsequent
     encounters to the same problem will be silent.

     This suppression only suppresses the error encountered on barfile.c,
     line 57.  All other asserts and notreached's are unaffected and still
     capable of generating full warnings and/or bombing out.  Multiple
     asserts can be suppressed at once, but you will only get the warning
     on the first one encountered.

DIAGNOSTICS

BUGS

SEE ALSO
     abort(3), setjmp(3)

AUTHOR
     Will Naylor, Bill Chapman


change_log.txt
/*******************************************************************************

5.0.1.20060413
- No changes, but now is release candidate.

20060325
- [RJL, all] Fixed more typos in the help docs.
- [CAF, linux] Updated wxWidgets patches and installation instructions.

20060323
- [RJL, all] Changed default symmetry angle for empty trees to 90 degrees.
- [RJL, all] Added discussion of color scheme to the View Settings window docs
and fixed a few typographical errors.
- [RJL, all] Changed RADIAN and DEGREES constants to be declared as constant
values rather than computed expressions, which fixed a problem in Intel Mac
builds (and only there) in which these constants were being given a value of 0.
- [RJL, all] Renamed Pi to PI, Two_Pi to TWO_PI for consistency and eliminated
const-ness of some return values for consistency.

5.0.0.20060309
- [RJL, all] Fixed a bug in the Path Combo inspector in which changing the
condition type wasn't always getting applied.
- [CAF, linux] - window fitting commands require WM hinting in some desktops;
make doc windows slightly less tall so bottom isn't hidden in some desktops.
- [CAF, all] Number pad delete key now deletes conditions in inspector (along
with backspace and fwd delete).
- [CAF, linux] Fixed missing double-click events in wxGTK list boxes
- [ABX, MSW] Printing dashing fixed.

20060308
- [CAF, Linux] Restored quit-on-close to match MSW behavior.

20060307
- [CAF, linux] invoke updating of user interface events in UpdateAllViews -
should patch the "unresponsive key after menu" issue in wxGTK
- [CAF, linux] like we did before in tmwxDocManager.cpp::OnFileClose (for closing
doc via menu/C-W), in tmwxDesignFrame::OnClose explicitly call base's method
instead of event.Skip(); [possibly a wxGTK, not Linux, issue]. Otherwise, when
a document window is killed via title bar button, the doc is not actually
closed, orphaning the inspector windows without creating a new empty doc.
- [RJL, all] Holding down modifier key when creating a new node by clicking
keeps the current selection.

20060304
- [ABX, MSW] Inspector and View Settings in on top when needed and on
background when unneeded. 
- [ABX, all] Added wrappers for range selection so it not influence existing
quality but is available for further development (just define
TM_WITH_RANGE_SELECTION macro in IDE/build). 
- [ABX, all] Assert during menu creation fixed.

20060228
- [RJL, all] Fixed a bug in tmConditionNodesCollinear in which deleting one of
the nodes could cause a crash.
- [RJL, all] Fixed an incorrect label in display of edges same strain condition.
- [RJL, all] In tmwxDesignCanvas::UpdatePaperSize, changed order of resizing
virtual image and resizing frame (frame now resized first) to avoid flicker
(that was only a problem on linux).

20060227
- [RJL, all] Rewrote Fit-to-[something] routines in tmwxDesignCanvas and
related supporting routines. Now supports user changes in monitor size and menu
items are enabled iff the command would alter the window position or size. Also
fixed several uninitialized members in tmwxDesignCanvas. Renamed several
tmwxDesignCanvas variables and routines.
- [RJL, all] Set paper size dialog is now initialized with current paper size.
- [RJL, all] Replaced a bunch of public member framework-related variables with
private members and public getters (to separate read/write access).
- [ABX, MSW+linux] Standard menu items are decorated with icons. (not on Mac.)
- [ABX, all] Drag & drop is supported.

20060226
- [RJL, all] Fixed a bug in the Node Fixed to Position dialog, in which checkbox
values were being ignored.
- [RJL, all] replacement of assorted ".size() == 0" with ".empty()" calls.
- [RJL, all] Eliminated superfluous arguments in tmwxTextCtrl::ValidateSizeT.
- [RJL, all] Fixed a bug in tmTree::AbsorbNode() in which mIsLeafNode and
mIsLeafPath flags were not getting updated properly in tmTree::AbsorbEdge.
- [ABX, all] Added check to Fit... commands so that they're not doubly executed
if the paper is already fit to the screen in one or the other direction.

20060224
- [ABX, all] Implemented View->Fit to Width and View->Fit to Height, added docs.
- [ABX, MSW] Added modified setup0.h (wxWidgets file) to turn off unnneeded
parts of wxW to reduce file size.
- [RJL, all] Replaced static arrays sCalcLocFns, sAddConditionFns, sSelectionFns
with singleton functions to delay initialization (resolving potential
intialization order issues). Also reimplemented debugging stringstreams with
initialize-on-usage.
- [RJL, all] Fixed an issue in node absorption that could cause a crash if a
crease pattern was present.

20060223
- [RJL, all] Fixed a bug in tmStubFinder::TriangulateTree in which encountering
of a non-splittable order-4+ tree (yes, it happens) caused a crash. Updated the
docs to note the possibility.
- [RJL, all] Fixed a bug in the display of a group selection (edges improperly
displayed). As part of that, implemented copy ctors and assignments in class
tmDpptrArray<T> to eliminate the underlying weakness (shallow default
ctor/assignment) that led to the bug.
- [RJL, all] Disabled Edit->Paste command (not used by any of our classes).
Framework still enables/handles for text controls.
- [ABX, all] Mouse event handler doesn't eat unused events to enable receiving
context menu event.
- [ABX, all] Context menu implemented; works both using keyboard context key
and right click from mouse. It just duplicates content of edit menu which is
handy in case of large space.

20060222
- [ABX, all] Fixed some typos and glitches in tmwxApp that were preventing
build on MSW.
- [RJL, all] Added some discussion of deletion (under Edit->Clear) to the help
docs.

20060221
- [CAF, linux] simplified code for small widgets; improved wxWidgets's circle
rendering for GNOME printing
- [RJL, all] Splash screen is now in its own routine, which happens at different
places in OnInit() depending on platform (later on Mac, earlier on Lin/Win).

Source/tmModel/wnlib/low/wnasrt.c
/*************************************************************************

wn_set_assert_print(passert_print)

wn_assert(cond)
wn_assert_notreached()
wn_assert_warn(cond)
wn_assert_warn_notreached()

*************************************************************************/
/****************************************************************************

COPYRIGHT NOTICE:

  The source code in this file is provided free of charge
  to the author's consulting clients.  It is in the
  public domain and therefore may be used by anybody for
  any purpose.

AUTHOR:

  Will Naylor, Bill Chapman

****************************************************************************/
#include <stdlib.h>
#include <stdio.h>
#include "wnlib.h"
#include "wnasrt.h"



local bool initialized = FALSE;

#ifdef __cplusplus
  extern "C" {
    local void (*psaved_assert_print)(const char string[]);
    local void (*psaved_assert_crash)(void);
  }
#else /* plain C */
  local void (*psaved_assert_print)(const char string[]);
  local void (*psaved_assert_crash)(void);
#endif


void wn_default_assert_print(const char string[])
{
  fprintf(stderr,"%s",string);
}


void wn_default_assert_crash(void)
{
  abort();
}


local void print_user_message(void)
{
  (*psaved_assert_print)(
      "------------------------------------------------------------\n");
  (*psaved_assert_print)(
      "                       HELP!!!\n");
  (*psaved_assert_print)(
      "This program has encountered a severe internal error.\n");
  (*psaved_assert_print)(
      "Please report the following message to the\n");
  (*psaved_assert_print)(
      "program's developers so that they can fix the problem,\n");
  (*psaved_assert_print)(
      "an override may be available:\n");
  (*psaved_assert_print)(
      "------------------------------------------------------------\n");
}


local void initialize_assert(void)
{
  if(!(initialized))
  {
    psaved_assert_print = &wn_default_assert_print;
    psaved_assert_crash = &wn_default_assert_crash;

    initialized = TRUE;
  }
}


EXTERN void wn_set_assert_print
(
  void (*passert_print)(const char string[])
)
{
  initialize_assert();

  psaved_assert_print = passert_print;
}


EXTERN void wn_set_assert_crash
(
  void (*passert_crash)(void)
)
{
  initialize_assert();

  psaved_assert_crash = passert_crash;
}


void wn_get_assert_override_string(char *output_buffer,
/**/          const char *file_name, int line_num)
{
  unsigned result;
  char *pc;
  int i, j;

  result = 0;
  for (i = j = 0;  file_name[i];  ++i, j = (j+1) % 4)
  {
    ((char *) &result)[j] += file_name[i];
  }

  result += line_num * line_num * line_num * line_num * line_num * line_num;

  sprintf(output_buffer, "WN_ASSERT_SUPPRESS_%s_%d_%x", file_name, line_num,
  /**/                result);
  for (pc = output_buffer;  *pc;  ++pc)
  {
    if ('.' == *pc)
    {
      *pc = '_';
    }
  }
} /* wn_get_assert_override_string */


/* Note that both func_name and/or exp_string might be NULL */

/*[RJL] added prototype */
void wn_general_assert_routine(const char *file_name, int line_num,
/**/  const char *func_name, const char *exp_string, bool warn_only,
/**/            bool notreached);
/*[\RJL] */
void wn_general_assert_routine(const char *file_name, int line_num,
/**/  const char *func_name, const char *exp_string, bool warn_only,
/**/            bool notreached)
{
  char string[1000];
  char override_string[1000];
  char *pc;

  /* if overridden, be absolutely silent, whether warning or fatal */
  wn_get_assert_override_string(override_string, file_name, line_num);
  if (getenv(override_string))
  {
    static bool first_assert_suppressed = TRUE;
    if (first_assert_suppressed)
    {
      (*psaved_assert_print)(
      "Warning: at least one assert suppressed, by environment variable\n");
      (*psaved_assert_print)("    \"");
      (*psaved_assert_print)(override_string);
      (*psaved_assert_print)("\"\n");
      first_assert_suppressed = FALSE;
    }

    return;
  }

  initialize_assert();

  print_user_message();

  if (notreached)
  {
    /* There is no expression, we just reached a notreached */

    (void)sprintf(string,
    /**/  "wn_assert%s_notreached() called -- %sforcing crash\n",
    /**/    warn_only ? "_warn" : "", warn_only ? "NOT " : "");
    (*psaved_assert_print)(string);
  }
  else
  {
    (void)sprintf(string, "Assertion botched -- %sforcing crash\n",
    /**/          warn_only ? "NOT " : "");
    (*psaved_assert_print)(string);

    /*     we want the expression on a separate line because it could be
    ** very long */
    if (exp_string)
    {
      (void) sprintf(string, "Expr (%s) was false\n", exp_string);
      (*psaved_assert_print)(string);
    }
  }

  pc = string;
  if (func_name)
  {
    (void) sprintf(string, "%cn %s() ", exp_string ? 'i' : 'I', func_name);
    for (  ;  *pc;  ++pc) ;
  }
  (void) sprintf(pc, "%ct line %d in file %s\n", func_name ? 'a' : 'A',
  /**/              line_num, file_name);
  (*psaved_assert_print)(string);

  if (!warn_only)
  {
    (*psaved_assert_crash)();
  }
} /* wn_general_assert_routine */


void wn_assert_routine(const char file_name[],int line_num)
{
  wn_general_assert_routine(file_name, line_num, NULL, NULL, FALSE, FALSE);
} /* wn_assert_routine */


void wn_assert_notreached_routine(const char *file_name, int line_num)
{
  wn_general_assert_routine(file_name, line_num, NULL, NULL, FALSE, TRUE);
}


void wn_assert_warn_routine(const char file_name[],int line_num)
{
  wn_general_assert_routine(file_name, line_num, NULL, NULL, TRUE, FALSE);
}


void wn_assert_warn_notreached_r(const char file_name[],int line_num)
{
  wn_general_assert_routine(file_name, line_num, NULL, NULL, TRUE, TRUE);
}


/* **************************************************************** */

void wn_assert_routine_func_exp(const char file_name[], int line_num,
/**/      const char *func_name, const char *exp_string)
{
  wn_general_assert_routine(file_name, line_num, func_name, exp_string,
  /**/                FALSE, FALSE);
} /* wn_assert_routine_func_exp */


void wn_assert_notreached_routine_func(const char *file_name, int line_num,
/**/            const char *func_name)
{
  wn_general_assert_routine(file_name, line_num, func_name, NULL,
  /**/                FALSE, TRUE);
}


void wn_assert_warn_routine_func_exp(const char file_name[], int line_num,
/**/      const char *func_name, const char *exp_string)
{
  wn_general_assert_routine(file_name, line_num, func_name, exp_string,
  /**/                TRUE, FALSE);
}


void wn_assert_warn_notreached_r_func(const char file_name[],int line_num,
/**/            const char *func_name)
{
  wn_general_assert_routine(file_name, line_num, func_name, NULL,
  /**/                TRUE, TRUE);
}
